keps/sig-node/2625-cpumanager-policies-thread-placement/README.md

```diff
@@ -0,0 +1,314 @@
+The static policy allocates CPUs using a topology-aware best-fit allocation. This enhancement wants to provide stronger guarantees by restricting the allocation of threads.
+
+Key properties:
+- Preserve all the properties of the `static` policy.
+- Never allocate less than a physical-cpu worth amount of cores.
+- With this requirement enforced, the cpumanager allocation algorithm will guarantee avoidance of physical core sharing.
+- Should the node not have enough free physical cores, the Pod will be put in Failed state, with `SMTAlignmentError` as reason.
+
+### Implementation strategy of smtaware CPU Manager policy
+
+- The CPU Manager implements the pod admit handler interface and participates in Kubelet pod admission.
+- The `GetAllocateResourcesPodAdmitHandler()` function in the Container Manager is modified to perform admission checks for CPU Manager in addition to the already occurring Topology manager admission check.
+Just like Topology Manager returns `TopologyAffinityError` is case of resources cannot be NUMA-aligned in case of `restricted` or `single-numa-node policy`, CPU Manager admission failure results in a rejected pod with `SMTAlignmentError`.
+- The Policy interface in CPU Manager is enhanced to support an Admit function. This allows to perform pod admission decision for CPU Manager policies.
+- When CPU Manager is configured with `smtaware` policy, when the `Admit()` function is called it is checked if the CPU request is
+such that it would acquire an entire physical core. In case request translates to partial occupancy of the cores the Pod will not be admitted and would fail with `SMTAlignmentError`. In case of `static` and `none` policy, the Admit() function always returns true meaning the pod is always admitted.
+- When `smtaware` policy is bootstrapped, we intentionally reuse `static` policy. The allocation logic of the `static` policy works seamlessly in case of `smtaware` policy as following the check at admission time of a pod, it is known that CPUs would be allocated such that full cores are allocated. Because of this check, a pod would never have to acquire single threads with the aim to fill partially-allocated cores.
+
+### Resource Accounting
+
+To illustrate the behaviour of the policy, we will consider the following CPU topology. We will use as example a CPU package with 16 physical cores, 2-way SMT-capable.
+
+![Example Topology](smtaware-topology.png)
+
+
+Let's consider a single container, requesting 5 isolated cores.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: frontend
+spec:
+  containers:
+  - name: app
+    image: images.my-company.example/app:v4
+    resources:
+      requests:
+        memory: "128Mi"
+        cpu: "5"
+      limits:
+        memory: "128Mi"
+        cpu: "5"
+```
+
+The `smtaware` policy would need to make sure the remaining core on the half-allocated physical CPU is left unallocated to avoid noisy neighbours.
```

Comment:
What is the idea about how this should be handled?

Reply:
Also, isn't the current way of making cores "exclusive" to only disallow other containers to use them? Is it enforced on other processes on the system as well.

Reply:
Yes, exclusive core allocation indeed prevents other workloads to run on the same cores, but on the same *virtual* cores.
A physical core with SMT enabled exposes 2 (typically) virtual cores, and these virtual cores share a fair amount of compute resources (actual sylicon).

Reply:
> Yes, exclusive core allocation indeed prevents other workloads to run on the same cores, but on the same virtual cores.

In "other workloads", do you mean other pods/containers in k8s, or _all_ workloads on the node? Not sure If I understand what you mean exactly...

I am just wondering how the intention is to enforce that no other workloads will run on one of the _virtual_ cores.

Reply:
I meant other workload managed by kubernetes (will doublecheck the KEP text to make sure this is clear).
Let me take the chance to summarize:

Let's put out of the picture processes outside kubernetes. Let's take for granted that appropriate node configuration (using for example systemd `CPUaffinity`, kubelet's `--reserved-cpus` option, proper kernel tuning) make sure that there is a set of cpus whose only manager is the kubernetes ' cpumanager. So in this scenario we want to have cpu isolation.

The static policy will exclusively allocate what kubernetes calls "cpus" to container, which will indeed ensure no other containers run on the same cpus. Since we work under assumption no workload on the node can possibly run on the same cpus, we are done!
Problem is: this is only one layer of isolation, there is still room for containers to cause noisy neighbours and to interfere to each other.

The next logical step we need to take is ensure isolation at core level. Here the terminolgy is a bit tricky because cpumanager calls "core" a "physical cpu" while kubernetes calls "cpu" a "virtual cpu" aka a hardware thread (https://kubernetes.io/docs/tasks/configure-pod-container/assign-cpu-resource/#cpu-units)
In this scenario (which we implement in the PR) we prevent two containers which gets virtual cpus aka hyperthread running on the same core aka physical cpu to interfere to each other. So a container only get full cores (physical cpus)

Admittedly this is an improvement for the noisy neighbours problem - not the full solution, for which we need to address L3 cache sharing, memory bandwidth and more. But this is a small and self contained enhancement which fits in the cpumanager.